src/libraries/System.Private.CoreLib/src/System/Globalization/InvariantModeCasing.cs

```diff
@@ -0,0 +1,369 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.Runtime.InteropServices;
+using System.Runtime.CompilerServices;
+using Internal.Runtime.CompilerServices;
+
+namespace System.Globalization
+{
+    internal static class InvariantModeCasing
+    {
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static char ToLower(char c) => CharUnicodeInfo.ToLower(c);
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static char ToUpper(char c) => CharUnicodeInfo.ToUpper(c);
+
+        internal static string ToLower(string s)
+        {
+            if (s.Length == 0)
+            {
+                return string.Empty;
+            }
+
+            ReadOnlySpan<char> source = s;
+
+            int i = 0;
+            while (i < s.Length)
+            {
+                if (char.IsHighSurrogate(source[i]) && i < s.Length - 1 && char.IsLowSurrogate(source[i + 1]))
+                {
+                    SurrogateCasing.ToLower(source[i], source[i + 1], out char h, out char l);
+                    if (source[i] != h || source[i + 1] != l)
```

Comment:
should you store `source[i]` and `source[i + 1]` in locals? It seems like they are used multiple times in this block.

Reply:
the input is a string which cannot change. We are moving it to local variable only when using span to avoid mutating the values in the middle.